### ISSUE TITLE
Extend the engine to run portable tests

### ISSUE THREAD

**Issue by rprouse** [nunit/nunit#1138]
This needs design, but could be platform specific agents. If so, it will likely be blocked by #362.

**Comment by rprouse** [comment]
In #1168, @CharliePoole said,

>I'm not sure there is a problem here since this refers (in my mind anyway) to an engine running on the desktop. What we need to find out is whether an engine running under .NET 2.0/3.5 is able to analyze assemblies built for other targets. I don't see why not, since they are not actually loaded. Consider that we are already analyzing .NET 4.5 assemblies. Of course, this needs to be tested on a VM that has only .NET 2.0 or 3.5 installed.

Based on my spikes into this so far, I think this is the way we need to go. The engine can inspect the test assembly and determine it's target platform. We just can't run it directly.

This is also mixed up with #677 and #362. My idea so far is as follows,

- The current NUnit3 engine inspects the test assembly in addition to the framework. It looks for the `TargetFrameworkAttribute` which will tell you for example if it is targeting `.NETCore,Version=v5.0`.
- If the target framework is a non-desktop framework, disallow `--inprocess` since we can't run non-desktop targets. We might also want to disallow `--process=Single` or we make sure all test assemblies target the same platform.
- Modify `TestAgency` to allow it to launch platform specific agents and communicate with them in a manner other than .NET remoting. This is #362. If we are running those agents on the desktop machine, we could even capture Console.In/Out for communications.

The platform specific agents are where decisions still need to be made. It would be nice to have a portable mini-engine (#677) that is capable of loading and running the portable framework. Like the full framework, this would allow us to run test no matter which version of the NUnit 3 framework is used. It would likely be a stripped down version of the engine that can only load from the agent directory and is only intended to be used by agents. This might even boil down to a portable `IFrameworkDriver` and some supporting classes?

A simplified and not mutually exclusive approach might be that the agents are actually the nunitlite version of the tests. We could modify nunitlite to run directly, or be able to communicate with the engine.

**Comment by CharliePoole** [comment]
I think we're pretty much on the same page at a high level. Some details...

* Can we count on TargetFrameworkAttribute being reliably set?

* We can't precisely disallow all those options, although we might be able to in some cases. Basically, we can only look at  explicit options, but if the defaults are used, we have to wait until we analyze the assembly and then reflect an error back to the runner.

* I think we still have two very distinct options for running portable and device-based tests: a scaled down engine versus a scaled up driver. Neil has been pushing us strongly toward an engine for CF but I'm not sure that's the  easiest approach. At some point, we have to spike both approaches. I think the easiest way to do that initially is right on the desktop itself.

Charlie

**Comment by rprouse** [comment]
>Can we count on TargetFrameworkAttribute being reliably set?

I have tested quite a few different assemblies and have not found it unset except for .NET 2.0 assemblies. I think that we can safely assume that if it is not set, it targets a full .NET framework. I assume that these days with so many potential targets, it needs to be set.

>I think we still have two very distinct options for running portable and device-based tests: a scaled down engine versus a scaled up driver. Neil has been pushing us strongly toward an engine for CF but I'm not sure that's the easiest approach. At some point, we have to spike both approaches. I think the easiest way to do that initially is right on the desktop itself.

I installed the [.NET Portability Analyzer](https://visualstudiogallery.msdn.microsoft.com/1177943e-cfb7-4822-a8a6-e56c7905292b) extension for Visual Studio and analyzed the engine against the platforms that we want to support. Based on what is not available, I think approaching it from the driver end is probably the best option. At a minimum, we need to be able to load the framework in a version independent way, explore or run the tests and communicate the results. The majority of the functionality of the engine is used in the startup process, not in the agents themselves.

**Comment by xied75** [comment]
:+1: to track.

**Comment by xied75** [comment]
:+1: to track.

**Comment by rprouse** [comment]
@xied75, for future reference, you can just click the subscribe button at the bottom of the toolbar on the left. That way, all other subscribers don't get an email :wink: 

![image](https://cloud.githubusercontent.com/assets/493828/13115455/ddafb478-d565-11e5-8eb0-becc60f5c3e9.png)

**Comment by xied75** [comment]
Well this is in fact a new trend on GitHub, since GitHub refuses to add the feature to allow one see what he/she subscribed, so now people do +1 in order to find what they want track.

**Comment by rprouse** [comment]
I am going to move this out of 3.4. The new `dotnet-test-nunit` will run .NET Core and UWP apps and the promise of `netstandard` may change the way that we approach this. I want to see how things go after the release of .NET Core before committing to major changes in the engine.

**Comment by CharliePoole** [comment]
That makes perfect sense.

**Comment by CharliePoole** [comment]
Issue moved to [nunit/nunit-console #10](https://github.com/nunit/nunit-console/issues/10) via [**ZenHub**](https://www.zenhub.com/)
